### PR TITLE
WFLY-6634 Add remote-cache-container configuration to Infinispan subsystem

### PR DESCRIPTION
**Pull request by rhusar**
…finispan subsystem

Jira
https://issues.jboss.org/browse/WFLY-6634

### CODE REVIEW

**Comment by jmesnil** [comment]
You'll need to create a PR against the EAP 7.x branch that references the corresponding JBEAP issue to get it merged.

**Comment by rhusar** [comment]
Hm, it's not clear yet if this is for EAP 7.1 or 7.2 so no downstream PR until then, but I will remove the [11.x] from the title, as that might be confusing. Unfortunately that's the highest number we have in Jira right now.

**Review by pferraro — CHANGES_REQUESTED** [review]

File: clustering/infinispan/extension/src/main/resources/schema/jboss-as-infinispan_5_0.xsd
Comment (pferraro) [inline clustering/infinispan/extension/src/main/resources/schema/jboss-as-infinispan_5_0.xsd]: You know my general philosophy on boolean attributes...
What do you think of replacing this with strategy="FIFO|LIFO" or pool="QUEUE|STACK"?

File: clustering/infinispan/extension/src/test/resources/org/jboss/as/clustering/infinispan/subsystem/subsystem-infinispan-5_0.xml
Comment (tristantarrant) [inline clustering/infinispan/extension/src/test/resources/org/jboss/as/clustering/infinispan/subsystem/subsystem-infinispan-5_0.xml]: Since we support failover clusters, the server configuration should be extracted to its own sub-element.
<remote-cache-container>
    <remote-servers>
       <remote-server socket-bindings="jdg1 jdg2 jdg3"/>
       <remote-server cluster-name="failoversite" socket-bindings="jdg4 jdg5 jdg6"/>
    </remote-servers>
</remote-cache-container>

Comment (pferraro) [inline clustering/infinispan/extension/src/test/resources/org/jboss/as/clustering/infinispan/subsystem/subsystem-infinispan-5_0.xml]: @tristantarrant Our XML parser is versioned - so future proofing is not an issue.

Comment (tristantarrant) [inline clustering/infinispan/extension/src/test/resources/org/jboss/as/clustering/infinispan/subsystem/subsystem-infinispan-5_0.xml]: Sure, ours is too, but I'd rather future proof the design now.

Comment (tristantarrant) [reply clustering/infinispan/extension/src/test/resources/org/jboss/as/clustering/infinispan/subsystem/subsystem-infinispan-5_0.xml]: My preference is for the schema to support a single <remote-server> initially with the possibility in the future to support multiples.
Probably instead of remote-server (which conveys the concept of a single node) this should be called remote-cluster

**Comment by jfclere** [comment]
@jmesnil what about the rebase-this flag, it looks good now.

**Comment by jmesnil** [comment]
@jfclere I removed the label

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/ConnectionPoolBuilder.java
Comment (pferraro) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/ConnectionPoolBuilder.java]: Use == for enums.

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/InfinispanSubsystemXMLWriter.java
Comment (pferraro) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/InfinispanSubsystemXMLWriter.java]: This is a required child, so this resource is guaranteed to exist.

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/InfinispanSubsystemXMLWriter.java
Comment (pferraro) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/InfinispanSubsystemXMLWriter.java]: Please delete these 3 lines.

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/InfinispanSubsystemXMLWriter.java
Comment (pferraro) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/InfinispanSubsystemXMLWriter.java]: There will always be at least 1 remote cluster, so this check is unnecessary.

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/InfinispanSubsystemXMLWriter.java
Comment (pferraro) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/InfinispanSubsystemXMLWriter.java]: This is a required child, so this resource is guaranteed to exist.

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/ThreadPoolDefinition.java
Comment (pferraro) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/ThreadPoolDefinition.java]: Why is this necessary?

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/ThreadPoolDefinition.java
Comment (pferraro) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/ThreadPoolDefinition.java]: Why is this necessary?

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/ClientThreadPoolResourceDefinition.java
Comment (pferraro) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/ClientThreadPoolResourceDefinition.java]: Why create another resource definition?  Why not just add a CLIENT value to the existing ThreadPoolResourceDefinition enum?

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/ClientThreadPoolBuilder.java
Comment (pferraro) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/ClientThreadPoolBuilder.java]: Can we not just reuse the existing ThreadPoolBuilder?

**Review by pferraro — CHANGES_REQUESTED** [review]

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/InfinispanSubsystemXMLWriter.java
Comment (rhusar) [reply clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/InfinispanSubsystemXMLWriter.java]: Fixed, removed.

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/InfinispanSubsystemXMLWriter.java
Comment (rhusar) [reply clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/InfinispanSubsystemXMLWriter.java]: Fixed, removed.

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/InfinispanSubsystemXMLWriter.java
Comment (rhusar) [reply clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/InfinispanSubsystemXMLWriter.java]: Oops, removed.

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/InfinispanSubsystemXMLWriter.java
Comment (rhusar) [reply clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/InfinispanSubsystemXMLWriter.java]: Right, the original idea was to support the collocated cluster. Removed now.

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/ThreadPoolDefinition.java
Comment (rhusar) [reply clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/ThreadPoolDefinition.java]: Fixed, removed.

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/ThreadPoolDefinition.java
Comment (rhusar) [reply clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/ThreadPoolDefinition.java]: Fixed, removed.

Comment (rhusar) [reply clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/ThreadPoolDefinition.java]: Originally needed to support both thread pool definitions in generic writer method.

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/ThreadPoolDefinition.java
Comment (rhusar) [reply clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/ThreadPoolDefinition.java]: Originally needed to support both thread pool definitions in generic writer method.

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/ClientThreadPoolResourceDefinition.java
Comment (rhusar) [reply clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/ClientThreadPoolResourceDefinition.java]: Merged into ThreadPoolResourceDefinition#CLIENT now.

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/ConnectionPoolBuilder.java
Comment (rhusar) [reply clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/ConnectionPoolBuilder.java]: Fixed.

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/ClientThreadPoolBuilder.java
Comment (rhusar) [reply clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/ClientThreadPoolBuilder.java]: This is needed because it builds diffent objects: org.infinispan.client.hotrod.configuration.ExecutorFactoryConfiguration as opposed to org.infinispan.configuration.global.ThreadPoolConfiguration....

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/ThreadPoolDefinition.java
Comment (pferraro) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/ThreadPoolDefinition.java]: ?

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerResourceDefinition.java
Comment (pferraro) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerResourceDefinition.java]: ThreadPoolResourceDefinition.CLIENT.register(registration);

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteClusterResourceDefinition.java
Comment (pferraro) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteClusterResourceDefinition.java]: Make sure to add this to wildfly-capabilities documentation.

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteConfigurationBuilder.java
Comment (pferraro) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteConfigurationBuilder.java]: You don't need to call getDefinition() - this method is exposed via the Attribute interface.

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerBuilder.java
Comment (pferraro) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerBuilder.java]: This should be ON_DEMAND.

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteConfigurationBuilder.java
Comment (pferraro) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteConfigurationBuilder.java]: This should be ON_DEMAND.

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/ClientThreadPoolBuilder.java
Comment (pferraro) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/ClientThreadPoolBuilder.java]: No need to extend DefaultAsyncExecutorFactory.

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/ClientThreadPoolBuilder.java
Comment (pferraro) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/ClientThreadPoolBuilder.java]: Just a TODO - figure out why Infinispan wants these to be daemon threads.

**Review by pferraro — CHANGES_REQUESTED** [review]

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteClusterResourceDefinition.java
Comment (rhusar) [reply clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteClusterResourceDefinition.java]: PR open https://github.com/wildfly/wildfly-capabilities/pull/20

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteConfigurationBuilder.java
Comment (rhusar) [reply clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteConfigurationBuilder.java]: Done, removed in all other places.

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteConfigurationBuilder.java
Comment (rhusar) [reply clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteConfigurationBuilder.java]: Done. Original when the PR was open we were still installing services as passive.

**Comment by tristantarrant** [comment]
As far as the schema is specified, I'm happy with it.

**Comment by rhusar** [comment]
Thanks Tristan!

**Review by pferraro — APPROVED** [review]

**Comment by rhusar** [comment]
@jmesnil This is ready to go for a while now, and since you are the only one with merge rights who have commented on the PR and also since you have indicated on other PR to hold on Infinispan 9 upgrade until Artemis upgrade, it looks like this PR should go first. Would you mind looking into it? Thanks!

**Comment by jmesnil** [comment]
@rhusar I mentioned to hold changes in messaging-activemq subsystem until Artemis 2.4 is released in #10678 but this PR was for upgrading JGroups, not Infinispan.
Afaik, there is no relation between Infinispan and the messaging-activemq subsystem so I have no issue with merging that PR.

**Comment by rhusar** [comment]
@jmesnil Just to clarify this: the Infinispan upgrade *depends* on the JGroups upgrade. So there is now a pretty long dependency chain which this PR is not a part of and can be merged in the meantime.

**Comment by jmesnil** [comment]
This PR was on hold until Artemis upgrade was done.
But this is taking longer than expected and I see no reason to hold this PR as @pferraro has approved the changes for it.

**Comment by jmesnil** [comment]
retest this please

**Comment by kabir** [comment]
@rhusar As this pull request has  been held up for a long time, I am closing it to get the pull request queue under control again. Since this is very old now, it will be a very long time until this one will be addressed. Also, I don't have much time to spend on this to evaluate on a case by case basis.

If this is something important to you or the project please feel free to recreate the pull request. That way it will come on top of the pile again, and be looked at again sooner. 

If it is a new **feature**, please follow the new process for features. Ping @kabir or @jamez if you are unsure what this means, and we can help. Please do that before opening the pull request though.

Sorry for any inconvenience!

**Comment by rhusar** [comment]
@kabir This is depressing.

**Comment by kabir** [comment]
I see this is a feature now :)

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerConfigurationBuilder.java
Comment (pferraro) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerConfigurationBuilder.java]: Additionally, we need to configure the Marshaller.  Similiar to what is done here:
https://github.com/wildfly-clustering/wildfly-clustering-web-hotrod/blob/master/src/main/java/org/wildfly/clustering/web/hotrod/HotRodMarshaller.java
Except that it should use a ModularClassResolver.

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerResourceDefinition.java
Comment (pferraro) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerResourceDefinition.java]: We need to add a "module" attribute used to configure the marshaller.

**Review by pferraro — CHANGES_REQUESTED** [review]

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerConfigurationBuilder.java
Comment (rhusar) [reply clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerConfigurationBuilder.java]: Done; added new commit to make review easier.

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerResourceDefinition.java
Comment (rhusar) [reply clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerResourceDefinition.java]: Done; added new commit to make review easier.

**Review by pferraro — APPROVED** [review]

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerResourceDefinition.java
Comment (pferraro) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerResourceDefinition.java]: Please replace these lambdas using the strategy employed here: https://github.com/wildfly/wildfly/pull/11079

**Review by pferraro — CHANGES_REQUESTED** [review]

**Review by pferraro — CHANGES_REQUESTED** [review]

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteClusterResourceDefinition.java
Comment (rhusar) [reply clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteClusterResourceDefinition.java]: Done.

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteClusterResourceDefinition.java
Comment (pferraro) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteClusterResourceDefinition.java]: looks like an extra space here?

Comment (rhusar) [reply clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteClusterResourceDefinition.java]: Well spotted, fixed thanks.

**Pull request merged**
